pkg/querier/block_streaming.go

```diff
@@ -181,41 +184,103 @@ func (s *storeGatewayStreamReader) StartBuffering() {
 		totalSeries := 0
 		totalChunks := 0
 
+		// Why do we buffer batches received before we receive the estimated number of chunks?
```

Comment:
The logic here is more complicated than I expected. I'm wondering why we can't keep the logic as is, and just increment the query limiter when the estimation is reached. So the PromQL engine starts processing the query, in the meanwhile we'll receive chunks estimations, and when the total chunks estimation hit the limit (if it will) we'll abort the execution. The difference with the actual chunks limit is that the estimations are expected be received sooner.

What am I missing?

Reply:
I think the main comment here is addressed by the change to always send the estimate before sending any chunks. 

The reason for enforcing the estimated limit early (before allowing the PromQL engine to start processing the query) is to mirror the behaviour of the non-estimated chunks limit, which is enforced before the PromQL engine begins evaluation for all cases except when streaming chunks from store-gateways.